contracts/02-tamagotchi-interaction/src/lib.rs

```diff
@@ -2,21 +2,138 @@
 
 #[allow(unused_imports)]
 use gstd::prelude::*;
+use gstd::{exec, msg, debug};
+use tamagotchi_interaction_io::{Tamagotchi, TmgAction};
 
-// TODO: 4️⃣ Define constants
+static mut TAMAGOTCHI_STATE: Option<Tamagotchi> = None;
+
+const HUNGER_PER_BLOCK: u64 = 1;
+const BOREDOM_PER_BLOCK: u64 = 2;
+const ENERGY_PER_BLOCK: u64 = 2;
+const FILL_PER_FEED: u64 = 1000;
+const FILL_PER_ENTERTAINMENT: u64 = 1000;
+const FILL_PER_SLEEP: u64 = 1000;
 
 #[no_mangle]
 extern fn init() {
-    // TODO: 0️⃣ Copy the `init` function from the previous lesson and push changes to the master branch
+    let name_bytes = msg::load_bytes().expect("Invalid initial name");
+    let name = String::from_utf8(name_bytes).expect("Invalid UTF-8");
+
+    let date_of_birth = exec::block_timestamp();
+
+    let tamagotchi = Tamagotchi {
+        name,
+        date_of_birth,
+        owner: msg::source(),
+        fed: 1000,
+        fed_block: 0,
+        entertained: 1000,
+        entertained_block: 0,
+        slept: 1000,
+        slept_block: 0,
+    };
+
+    save_tamagotchi_state(tamagotchi);
+
+    debug!("Tamagotchi initialized with name: {:?}, date of birth: {:?}", name, date_of_birth);
 }
 
 #[no_mangle]
 extern fn handle() {
-    // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
-    // TODO: 5️⃣ Add new logic for calculating the `fed`, `entertained` and `slept` levels
+    let action: TmgAction = msg::load().expect("Failed to decode TmgAction");
+
+    let current_block_height = exec::block_height() as u64;
+    let tamagotchi = unsafe {
+        TAMAGOTCHI_STATE
+            .as_mut()
+            .expect("The contract is not initialized")
+    };
+
+    update_levels(tamagotchi, current_block_height);
```

Comment:
You check the levels of the Tamagotchi and also send messages that the Tamagotchi is hungry, bored, or tired. The task stated that if a Tamagotchi is hungry, bored or tired, this means that the Tamagotchi is dead and there is no point in feeding it or playing with it. However, after this function you continue executing the message.
Also, when performing one or another action (Feed, entertain or sleep), you do not send an appropriate reply `TmgEvent`